### PR TITLE
Avoid leaking watch goroutines

### PR DESCRIPTION
**Pull request by wojtek-t**
Kind-of follow up to https://github.com/kubernetes/kubernetes/pull/84693

```release-note
NONE
```

### CODE REVIEW

**Comment by liggitt** [comment]
unit test failure looks relevant

**Comment by answer1991** [comment]
Watch request handler has so much long call stack, even dose not sure when and where we should close `watcher` :-P

I also have fixed a `watcher.Stop` before: https://github.com/kubernetes/kubernetes/pull/85410

**Comment by answer1991** [comment]
Duplicated with https://github.com/kubernetes/kubernetes/pull/85410.

1. Suggest call `watcher.Stop()` at `serveWatch`, as `serveWatch` may exit before serve watching.
2. Actually it's not leak a goruntine, as the real `watcher` is context-aware, see: https://github.com/kubernetes/kubernetes/pull/84693#issuecomment-555817112

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go [hunk 1]
```diff
@@ -172,6 +172,12 @@ func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 		return
 	}
 
+	// ensure the connection times out
+	timeoutCh, cleanup := s.TimeoutFactory.TimeoutCh()
+	defer cleanup()
+	// Ensure that underlying watch is stopped independently of any errors.
+	defer s.Watching.Stop()
```
Comment (answer1991) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go]: `serveWatch` may exit when occur an error before `ServeHTTP`.

**Comment by wojtek-t** [comment]
Closing in favor of https://github.com/kubernetes/kubernetes/pull/85410

**Pull request closed without merging**